### PR TITLE
#158903255 Change The Relationship Between Asset Category and Attribute To Many To Many

### PR DESCRIPTION
**Pull request by SEUNAGBEYE**
#### Description of Task to be completed?
- change the relationship from `one to many` to `many to many`
- rename `attribute` table to `attributes`
#### Screenshots (if appropriate)
N/A

### CODE REVIEW

File: api/models/asset_category_attribute_association.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+Asset Category and Attribute Relationship Module
+"""
+
+from .database import db
+
+asset_category_attribute_assoc = db.Table(
+    'asset_categories_attributes',
+    db.Column('asset_category_id', db.String, db.ForeignKey('asset_categories.id')),  # noqa
+    db.Column('attribute_id', db.String, db.ForeignKey('attribute.id')),
+)
```
Comment (GbengaOshinaga) [inline api/models/asset_category_attribute_association.py]: @SEUNAGBEYE Add newline here

File: tests/test_user_model.py [hunk 2]
```diff
@@ -1,29 +1,68 @@
...
+        assert user_query.filter(new_user.name == 'Ayobami').count() == 1
+        assert isinstance(user_query.filter(new_user.name == 'Ayobami').all(), list)  # noqa
 
-  def test_count(self, new_user):
-    assert new_user.count()  == 1 
+    def test_delete(self, new_user, request_ctx, mock_request_obj_decoded_token):  # noqa
+        """
+        Should delete a user
+        """
+        new_user.delete()
```
Comment (GbengaOshinaga) [inline tests/test_user_model.py]: @SEUNAGBEYE Can you add an assertion in this test case, to signify that the operation succeeded?

Comment (SEUNAGBEYE) [reply tests/test_user_model.py]: @GbengaOshinaga Done.

File: api/models/asset_category_attribute_association.py (hunk 1 above)
Comment (SEUNAGBEYE) [reply api/models/asset_category_attribute_association.py]: @GbengaOshinaga Done

File: tests/test_user_model.py [hunk 3]
```diff
@@ -1,29 +1,94 @@
...
+        user = User.get_or_create_instance({'id': 'wrong id'})
+        assert isinstance(user, User)
+
+    def test_get_or_create_instance_two(self, new_user):
+        """
+        Should get an instance of user
+        """
+        user = {}
+        user.update(new_user.__dict__)
+        del user['_sa_instance_state']
```
Comment (SEUNAGBEYE) [inline tests/test_user_model.py]: Add inline comments to line `75 & 76`

File: tests/test_user_model.py [hunk 4]
```diff
@@ -1,29 +1,94 @@
...
+            User.get_or_404('wrong id')
+
+    def test_get_or_404_success(self, new_user):
+        """
+        Should successfully get a user
+        """
+        user = User.get_or_404('1')
+        assert user == new_user
+
+    def test_get_or_create_instance(self):
```
Comment (SEUNAGBEYE) [inline tests/test_user_model.py]: Change method name to something more descriptive

File: tests/test_user_model.py [hunk 5]
```diff
@@ -1,29 +1,94 @@
...
+        user = User.get_or_create_instance(user)
+        assert isinstance(user, User)
+
+    def test_get_or_create_instance_error(self, new_user):
+        """
+        Should get an instance of user
+        """
+        try:
+            user = User.get_or_create_instance(new_user.__dict__)
+        except Exception as error:
```
Comment (SEUNAGBEYE) [inline tests/test_user_model.py]: Test if the Exception being thrown is catch or not

**Comment by SEUNAGBEYE** [comment]
@ibutiti @Dheavyman Please check the changes made.

File: api/models/asset_category_attribute_association.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+"""
+Asset Category and Attribute Relationship Module
+"""
+
+from .database import db
+
+asset_category_attribute_assoc = db.Table(
+    'asset_categories_attributes',
+    db.Column('asset_category_id', db.String, db.ForeignKey('asset_categories.id')),  # noqa
```
Comment (ibutiti) [inline api/models/asset_category_attribute_association.py]: You can put each of the args in a new line here and remove the `#noqa`. Same to line 10

File: api/models/model_operations.py [hunk 7]
```diff
@@ -47,6 +56,24 @@ def get_or_404(cls, id):
...
+    def get_or_create_instance(cls, kwargs):
+        """
+        return entries by id
+        """
+        from sqlalchemy.exc import InvalidRequestError
+        try:
+            record = cls.query.filter_by(**kwargs).first()
+        except InvalidRequestError:
+            raise ValidationError(
+                {'message': '_sa_instance should not be part of the dict'}
```
Comment (ibutiti) [inline api/models/model_operations.py]: Just wondered, will the front end know what `_sa_instance` is and why it came about? Can we add in the reason this key is in the object?

File: tests/test_user_model.py [hunk 8]
```diff
@@ -1,29 +1,100 @@
+"""Module for User Model Test"""
+
 from flask import jsonify
+import pytest
```
Comment (ibutiti) [inline tests/test_user_model.py]: Full imports should come before scoped imports.

File: tests/test_user_model.py [hunk 9]
```diff
@@ -1,29 +1,100 @@
...
+
+    def test_get_or_create_instance_to_create_an_instance(self):
+        """
+        Should create an instance of user if record does not exists
+        """
+        user = User.get_or_create_instance({'id': 'wrong id'})
+        assert isinstance(user, User)
+
+    def test_get_or_create_instance_to_get_an_instance(self, new_user):
+        """
```
Comment (frankip) [inline tests/test_user_model.py]: nice work on the comments now we can understand what the test does

File: api/models/model_operations.py (hunk 7 above)
Comment (SEUNAGBEYE) [reply api/models/model_operations.py]: @ibutiti This error is not thrown due to client bad request: it is thrown when a `backend dev` provide `invalid` data to the `filter_by` if the implementation is wrong.

**Comment by SEUNAGBEYE** [comment]
@brayoh This PR is ready to be merged

**Pull request closed without merging**